### PR TITLE
✨ Bring your own (encryption) key (BYOK)

### PR DESCRIPTION
**Pull request by akutz**
**What does this PR do, and why is it needed?**

This patch adds support for bringing your own encryption key used to encrypt/recrypt VMs.

**Which issue(s) is/are addressed by this PR?** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes `NA`

**Are there any special notes for your reviewer**:

This PR will be rebased once #670 and #672 are merged.

**Please add a release note if necessary**:

```release-note
Support bringing your own (encryption) key (BYOK) to encrypt VM workloads
```

----
📚 Documentation preview 📚: https://vm-operator--673.org.readthedocs.build/en/673/

### CODE REVIEW

File: pkg/providers/vsphere/vmprovider_vm.go
Comment (bryanv) [inline pkg/providers/vsphere/vmprovider_vm.go:1120]: I think this result in a Reconfigure() on every reconcile

File: api/v1alpha3/virtualmachine_types.go
Comment (rtsanev-github) [inline api/v1alpha3/virtualmachine_types.go:427]: if a default Key Provider is defined in the underlying platform...

It does not necessarily has to be native key provider, it can be standard key provider but it must be set as default in vCenter.

File: api/v1alpha3/virtualmachine_types.go
Comment (rtsanev-github) [inline api/v1alpha3/virtualmachine_types.go:431]: NKP > Default Key Provider 

here and below

File: api/v1alpha3/virtualmachine_types.go
Comment (rtsanev-github) [inline api/v1alpha3/virtualmachine_types.go:432]: Maybe its worth noting somewhere that in case there is a default key provider in vCenter AND EncryptionClassName (ECN) is specified, then ECN takes precedence.

File: test/builder/vcsim_test_context.go
Comment (dougm) [inline test/builder/vcsim_test_context.go:739]: Use the exported const `pbm/simulator.DefaultEncryptionProfileID` instead ?
We also could/should add one for ID used in `StorageProfileID` above.

**Review by dougm — APPROVED** [review]
lgtm, just a few minor comments/suggestions

File: pkg/providers/vsphere/session/session_vm_update.go
Comment (akutz) [reply pkg/providers/vsphere/session/session_vm_update.go:1228]: Ha, yeah, you and @bryanv literally had the exact same thought process :)

File: test/builder/vcsim_test_context.go
Comment (akutz) [reply test/builder/vcsim_test_context.go:739]: Yep. It did not exist when I wrote this. I'll switch to that :)

File: controllers/storageclass/storageclass_controller.go
Comment (akutz) [reply controllers/storageclass/storageclass_controller.go:91]: Yes, because here I'd rather not use reflection every time an object is reconciled. If I need to rebase this again, I'll add the reflected name to the reconciler struct and re-use it here.

File: api/v1alpha3/virtualmachine_types.go
Comment (aruneshpa) [inline api/v1alpha3/virtualmachine_types.go:386]: What are your thoughts on calling this `encryptionClassName`?  There's no stuttering since the parent field is called `spec.crypto`.   IMHO, `spec.cryto.encryptionClassName` is more accurate than `spec.crypto.className`.

Not to mention we have another field called `className` which might be confusing.

File: api/v1alpha3/virtualmachine_types.go
Comment (aruneshpa) [inline api/v1alpha3/virtualmachine_types.go:376]: Hmm.  I think this has to be more nuanced than that.

I think by "VM's home files", you mean files in VM_HOME directory.  But that also include PVC disks (FCDs can be created in either the designated fcd/ directory, or VM_HOME).

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:386]: Let me think about it. I am leaning towards agreeing with you.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:376]: No, by VM home files I mean any files in the VM directory that are *not* disks.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go]: Done, thanks @aruneshpa !

File: api/v1alpha3/virtualmachine_types.go
Comment (dilyar85) [inline api/v1alpha3/virtualmachine_types.go:379]: Nit: extra "and a" here.

File: controllers/storageclass/storageclass_controller.go
Comment (dilyar85) [inline controllers/storageclass/storageclass_controller.go:63]: "VirtualMachineClass": should this be "StorageClass" based on `controlledType = &storagev1.StorageClass{}`  above?

File: controllers/storageclass/storageclass_controller.go
Comment (dilyar85) [inline controllers/storageclass/storageclass_controller.go:92]: Nit: move the `logger/ctx` update after the "if" statement since they aren't used until then?

File: controllers/virtualmachine/virtualmachine/virtualmachine_controller.go
Comment (dilyar85) [inline controllers/virtualmachine/virtualmachine/virtualmachine_controller.go:221]: Also use `logger.V(4)` here to avoid log spam (since this will be called regardless of any requests exist)?

File: docs/concepts/workloads/vm.md
Comment (dilyar85) [inline docs/concepts/workloads/vm.md:565]: extra "then"

File: docs/concepts/workloads/vm.md
Comment (dilyar85) [inline docs/concepts/workloads/vm.md:540]: I noticed you have a mixed use of "encryption storage class" and "encrypted storage class" here and also in the code. Should we stick to one term ("encryption storage class") to keep it consistent with the API field?

File: webhooks/virtualmachine/validation/virtualmachine_validator.go
Comment (dilyar85) [inline webhooks/virtualmachine/validation/virtualmachine_validator.go:496]: This would allow users to specify `spec.crypto` fields (other than `encryptionClassName`) when the FSS is not enabled. Consider returning early if the `spec.crypto` is empty as well?

File: webhooks/virtualmachine/validation/virtualmachine_validator.go
Comment (dilyar85) [inline webhooks/virtualmachine/validation/virtualmachine_validator.go:485]: Should it be `encryptionClassName` at the end? This is not captured in test since you hard-coded the field path with this incorrect name.

**Review by dilyar85 — COMMENTED** [review]
Dropping the first round of review for the API, controller and webhook code. Will post another review for the rest later.

File: pkg/util/kube/storage.go
Comment (aruneshpa) [inline pkg/util/kube/storage.go:126]: Here and below, simplify with:

`return false, client.IgnoreNotFound(err)`

File: webhooks/virtualmachine/validation/virtualmachine_validator.go
Comment (akutz) [reply webhooks/virtualmachine/validation/virtualmachine_validator.go:496]: Yes, because `crypto.useDefaultKeyProvider` is currently the default behavior for today. But yes, I should disallow setting that field to a non-nil value as well.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:412]: Hmm, fair. However, I wrote the API docs from the perspective of the reconciler, not the webhooks (create vs update). Let me think on this.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:415]: Because if the underlying key or provider is removed, and the VM is not rekeyed to use the default provider, then you cannot power on the VM since it is encrypted and there is no key to decrypt it.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:769]: Hmmm. True. I will likely update the status to indicate that `Disks` does not include PVCs.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:768]: It does. But the documentation explains this (see `vm.md`). VMFiles is not descriptive either unless you already know.

File: pkg/util/kube/storage.go
Comment (akutz) [reply pkg/util/kube/storage.go:39]: Because nothing stops someone from modifying the underlying Storage Policy to no longer support encryption. This function is called by the StorageClass controller that ensures we always represent the current state of the underlying policy.

File: pkg/util/kube/storage.go
Comment (akutz) [reply pkg/util/kube/storage.go:65]: I'm keeping it, and ack.

File: pkg/util/kube/storage.go
Comment (akutz) [reply pkg/util/kube/storage.go:74]: I am using OwnerRefs because Sandeep will not allow me to annotate the StorageClasses and he did not like me placing a finalizer on them either when I was directly modifying the ConfigMap with the SC names. Neither the GoDocs nor the Kube docs mention any limit.

File: pkg/util/kube/storage.go
Comment (akutz) [reply pkg/util/kube/storage.go:98]: I wrote CreateAndPatch, so I am aware of what it does. The model I am using here is more efficient since I do not need to ever update *or* create the object if `encryption` is false.

File: docs/concepts/workloads/vm.md
Comment (akutz) [reply docs/concepts/workloads/vm.md:565]: No, the extra then is a typo for `the` :)

**Review by dilyar85 — APPROVED** [review]
LGTM with some minor comments left above. Thanks @akutz for the clear docs and comments on this feature!

File: controllers/storageclass/storageclass_controller.go
Comment (akutz) [reply controllers/storageclass/storageclass_controller.go:97]: That is correct. Any errors will be reported via controller-runtime. I do not believe in happy path logging except where it provides value. Here it does nothing but clutter logs.

File: pkg/util/kube/storage.go
Comment (akutz) [reply pkg/util/kube/storage.go:48]: Then you did not keep reading :) It is used for testing, hence the `internal` package. That way I can use it between prod and test without exposing it to beyond these two packages.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:770]: I updated the GoDocs/OpenAPI docs as well to explain this further.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:416]: I updated the comment to make it more clear.

File: pkg/vmconfig/vmconfig_reconciler.go
Comment (aruneshpa) [inline pkg/vmconfig/vmconfig_reconciler.go:19]: I know you are trying to introduce a pattern for this.  But reading through this code and the diff in the PR, my opinion is that we introduce this pattern when we have a _need_ for it.  Right now, it seems a bit too complicated for something that might not even be used by other components (if we never get to breaking the current pre/post power on reconfigure thing in this mode).

If you don't mind, I would recommend shelving all this, keeping this change specific to reconciling the StorageClasses and the VMs that are referencing them.

File: pkg/vmconfig/crypto/crypto_reconciler_pre.go
Comment (aruneshpa) [inline pkg/vmconfig/crypto/crypto_reconciler_pre.go:28]: Hmm.  Is this a pattern?  What does this buy us compared to just using `*vimtypes.VirtualMachineConfigSpec`?

File: pkg/vmconfig/crypto/crypto_reconciler.go
Comment (aruneshpa) [inline pkg/vmconfig/crypto/crypto_reconciler.go:95]: Andrew, I would rather have us just use the reasons explicitly, as opposed to define a type for them, and then use bitmasks to figure out the reasons that apply.  This will be hard to maintain and I don't see many usecases for this pattern. 

My recommendation would be to shelve the bitmask manipulation entirely and references the reasons explicitly where needed.

Comment (akutz) [reply pkg/vmconfig/crypto/crypto_reconciler.go:95]: The reasons *are* used explicitly, but there can be many different kinds and combinations. So I disagree there is not a use case for this in the future. There is literally a use case for it in this PR. The documentation in `vm.md` clearly outlines the reasons to the user. I appreciate the feedback, but I am going to keep this as-is.

File: pkg/vmconfig/crypto/crypto_reconciler_pre.go
Comment (akutz) [reply pkg/vmconfig/crypto/crypto_reconciler_pre.go:28]: It is used elsewhere in our codebase, yes. The types are exactly the same, and it enables more concise code. Even intellisense shows what it actually is.

File: pkg/vmconfig/vmconfig_reconciler.go
Comment (akutz) [reply pkg/vmconfig/vmconfig_reconciler.go:19]: I appreciate the feedback, but I am going to keep this. It is part of my vision going forward, and this change highlights *that* the pattern works and *how* the pattern works. I initially did this change without the pattern, and it ended up looking very similar, just without the wrapper. All of VM Op should start adopting this pattern in order to make our reconciler less fragile than it is and so difficult to change.

File: api/v1alpha3/virtualmachine_types.go
Comment (akutz) [reply api/v1alpha3/virtualmachine_types.go:413]: So we cannot avoid this situation. If the underlying key provider is gone, then a user has to be able to set `spec.crypto.encryptionClassName` to empty to bypass the failure state of being unable to find the `EncryptionClass`. When they set it to empty, the VM is either rekeyed using the default provider or not, based on the value of `spec.crypto.useDefaultKeyProvider`. This is aligned with the existing behavior/options of VCD.

File: pkg/vmconfig/crypto/crypto_reconciler.go
Comment (akutz) [reply pkg/vmconfig/crypto/crypto_reconciler.go:95]: FWIW, here is [a test case](https://github.com/vmware-tanzu/vm-operator/pull/673/files#diff-916fea486b6a687a518cffddb850c48ae19a10f9f42864ac3e3507126a6cc233R324-R341) that shows when the reason can be both `InvalidState` _and_ `InvalidChanges`

Comment (akutz) [reply pkg/vmconfig/crypto/crypto_reconciler.go:95]: All that said, this _is_ something I am willing to revisit. I will give it more thought, but I am not going to let it block this PR. After it merges, I'll sit down and consider how many possible combinations there can be in practical terms. If it is just `InvalidState` and `InvalidChanges`, then I can have a distinct reason for that. Let me look.

**Comment by akutz** [comment]
@aruneshpa I have addressed all of your feedback on the PR. While I did not accept it all, and stated my reasons, unless you have a strong objection, I am going to merge this PR so it does not continue to grow. We can revisit your concerns post-merge if that is alright?

**Comment by aruneshpa** [comment]
> @aruneshpa I have addressed all of your feedback on the PR. While I did not accept it all, and stated my reasons, unless you have a strong objection, I am going to merge this PR so it does not continue to grow. We can revisit your concerns post-merge if that is alright?

Sounds good to me.  Thanks!

**Pull request merged**
